Sources/MongoSwift/ConnectionPool.swift

```diff
@@ -67,38 +93,80 @@ internal class ConnectionPool {
         }
     }
 
-    /// Closes the pool, cleaning up underlying resources. This method blocks as it sends `endSessions` to the server.
-    internal func shutdown() {
-        switch self.state {
-        case let .open(pool):
-            mongoc_client_pool_destroy(pool)
-        case .closed:
-            return
+    /// Closes the pool, cleaning up underlying resources. **This method blocks until all connections are returned to
+    /// the pool.**
+    internal func close() throws {
+        try self.stateLock.withLock {
+            switch self.state {
+            case let .open(pool):
+                self.state = .closing(pool: pool)
+            case .closing, .closed:
+                throw Self.PoolClosedError
+            }
+        }
+
+        // continually loop and wait to get all connections back before destroying the pool. release the lock on each
+        // iteration to allow other methods to acquire the lock.
+        var done = false
+        while !done {
```

Comment:
Rather than using what is essentially a spin lock, it may be better to use something that can be signaled like a [condition variable](https://developer.apple.com/documentation/foundation/nscondition):
```swift
self.stateLock.lock()
while self.connCount > 0 {
    self.stateLock.wait() // waiting on a condition variable unlocks it
}
defer { self.stateLock.unlock() } // once the condition is met we hold the lock again
switch self.state {
    // do stuff
}
```
And in `checkIn:`
```swift
self.stateLock.lock()
defer { self.stateLock.unlock() }
...
self.connCount -= 1
self.stateLock.signal() // signal that the condition has changed, waking up any blocked `close`s
...
```
There's also [`NSConditionLock`](https://developer.apple.com/documentation/foundation/nsconditionlock) which is similar, but I don't think it would work for our case because we don't want a condition to be required for something to acquire the lock anywhere outside of `close`.

Reply:
maybe I'm misunderstanding your proposal, but I think we end up with a deadlock situation here:
```swift
self.stateLock.lock()
while self.connCount > 0 {
    self.stateLock.wait() // waiting on a condition variable unlocks it
}
```
we hold the lock and are waiting for `connCount` to drop to 0, but `close` needs to get the lock itself in order to update `connCount`.

Reply:
aha, I missed this from the docs:
> When a thread waits on a condition, the condition object unlocks its lock and blocks the thread. When the condition is signaled, the system wakes up the thread. The condition object then reacquires its lock before returning from the wait() or wait(until:) method. Thus, from the point of view of the thread, it is as if it always held the lock.

Reply:
`wait` unlocks the lock until it's woken up. The condition is then checked and if it doesn't pass, the `wait` is entered again and the lock is released again. If the condition passes, the lock will be held until the `defer` executes.

From the docs for `NSCondition`:
>When a thread waits on a condition, the condition object unlocks its lock and blocks the thread. When the condition is signaled, the system wakes up the thread. The condition object then reacquires its lock before returning from the wait() or wait(until:) method. Thus, from the point of view of the thread, it is as if it always held the lock.

With a normal lock, this is hard to implement without doing busy waiting like we're doing here or sleeping + periodic checks.

Reply:
jinx :)